pkg/executor/aggregate/agg_hash_final_worker.go

```diff
@@ -176,17 +242,60 @@ func (w *HashAggFinalWorker) receiveFinalResultHolder() (*chunk.Chunk, bool) {
 
 func (w *HashAggFinalWorker) run(ctx sessionctx.Context, waitGroup *sync.WaitGroup) {
 	start := time.Now()
-	defer func() {
-		if r := recover(); r != nil {
-			recoveryHashAgg(w.outputCh, r)
+	defer w.cleanup(start, waitGroup)
+
+	failpoint.Inject("enableAggSpillIntest", func(val failpoint.Value) {
+		if val.(bool) {
+			intestBeforeFinalWorkerStart()
 		}
-		if w.stats != nil {
-			w.stats.WorkerTime += int64(time.Since(start))
+	})
+
+	if w.spillHelper.isSpilledChunksIOEmpty() {
+		err := w.consumeIntermData(ctx)
+		if err != nil {
+			w.outputCh <- &AfFinalResult{err: err}
+			return
+		}
+	} else {
+		if w.spillHelper.checkError() {
+			return
 		}
-		waitGroup.Done()
-	}()
-	if err := w.consumeIntermData(ctx); err != nil {
-		w.outputCh <- &AfFinalResult{err: err}
 	}
-	w.loadFinalResult(ctx)
+	w.sendFinalResult(ctx)
+}
+
+func (w *HashAggFinalWorker) cleanup(start time.Time, waitGroup *sync.WaitGroup) {
+	if r := recover(); r != nil {
+		recoveryHashAgg(w.outputCh, r)
+	}
+	if w.stats != nil {
+		w.stats.WorkerTime += int64(time.Since(start))
+	}
```

Comment:
extract an individual func for increaseWorkerTime, there are dup codes in the partialWorker.

Reply:
> extract an individual func for increaseWorkerTime, there are dup codes in the partialWorker.

done